news/_posts/2019-15-01-resilient-php-applications.md

```diff
@@ -0,0 +1,108 @@
+* after a defined amount of time (called "reset timeout" sometimes), the Circuit Breaker try to access the service: if the service is available it will be used again, else the alternative response is provided instead;
+
+![https://martinfowler.com/bliki/images/circuitBreaker/sketch.png](https://martinfowler.com/bliki/images/circuitBreaker/sketch.png)
+
+> Schema from Martin Fowler's blog.
+
+This is also a big change of mindset for both the e-merchants and the developers, because for each process they need to think about what should be done in case of unavailability of a service.
+
+Like we already do in some old web browsers for the design of our websites, we should think about degraded experience when some third party services we use are not availables.
+
+### How works the PrestaShop Circuit Breaker?
+
+We were not satisfied with the quality and the constraints of the existing libraries or Symfony bundles:
+
+* we want something **really** extensible;
+* we want something **fully tested** and with **high quality** standards;
+* we want something that doesn't require uncommon PHP extensions (like *pcntl*)
+* and ... we still need to provide PHP 5.6+ compatibility;
+
+This is why we have worked on our own implementation of the design pattern, making adapters that relies on the popular HTTP Client [Guzzle](http://docs.guzzlephp.org/en/stable/index.html) to make calls and on the [Symfony Cache](https://symfony.com/doc/3.4/components/cache.html#simple-caching-psr-16) component for the storage/logging of the calls.
+
+Let's see how it will be used in PrestaShop modules controllers starting from PrestaShop 1.7.6:
+
+```php
+$fallbackResponse = function () {
+    return '{
+        "error": "service is unreachable right now."
+    }';
+};
+
+$this->call('https://api.some-domain.com', $fallbackResponse);
+```
+
+As this library doesn't depends on PrestaShop core libraries, you can use it in every PHP project:
+
+```php
+use PrestaShop\CircuitBreaker\SimpleCircuitBreakerFactory;
+
+$circuitBreakerFactory = new SimpleCircuitBreakerFactory();
+$circuitBreaker = $circuitBreakerFactory->create(
+    [
+        'closed' => [2, 0.1, 0],
+        'open' => [0, 0, 10],
+        'half_open' => [1, 0.2, 0],
+        'client' => ['proxy' => '192.168.16.1:10'],
+    ]
+);
+
+$fallbackResponse = function () {
+    return '{}';
+};
+
+$circuitBreaker->call('https://api.domain.com', $fallbackResponse);
+```
+
+## What’s next?
+
+We will start to use it on the "so well-known" _gamification_ module soon,
+and the library will be available in PrestaShop in the next minor release (1.7.6). Step by step, we will make PrestaShop Core and native modules resilient to networks failures.
+
+As most of our work, we have released the circuit breaker under MIT license and on our main PrestaShop organization. You are very welcome to contribute the library and to fill issues if you have specific needs.
```

Comment:
Most of PrestaShop's developers work is actually released under OSL and AFL.

"The Circuit Breaker library is developed and distributed under the MIT license, like many other librairies and dependencies included in the PrestaShop project (example: Symfony). This way, it is possible for anyone to use it with any other project. Your feedback and your contribution are welcome: feel free to open pull requests and issues about your needs.